### PR TITLE
Add Llama 3.1 8B QLoRA TT-Crank Experiments

### LINKED ISSUE #766: Add Llama 3.1 8B QLoRA TT-Crank Experiments
Add Llama 3.1 8B QLoRA experiments through tt-crank using Bfp.Bf8 and Bfp.Bf4 as weight storage dtypes.

### PR DESCRIPTION
**Pull request by tstepanovicTT**
### Issue
Closes [#766 ](https://github.com/tenstorrent/tt-blacksmith/issues/766)

### Training Workload Setting

- **Model:** Llama 3.1 8B
- **Strategy:** QLoRA
- **Framework:** PyTorch
- **Task:** Sentiment prediction, Instruction following
- **Dataset:** SST-2, Alpaca
- **Hardware:** P150

The training runs through tt-blacksmith with the tt-mlir env and uses `train_crank.py` as the equivalent of existing tt-blacksmith's XLA `train.py`. Until issue [#56613](https://github.com/tenstorrent/tt-metal/issues/56613) on tt-metal is solved, use `export TT_METAL_PINNED_MEMORY_CACHE_LIMIT_BYTES=536870912` to avoid hanging. 

### What's Changed
This PR showcases training workload for Llama 3.1 8B QLoRA fine-tuning on SST-2 and Alpaca through tt-crank, using Bfp.Bf8 and Bfp.Bf4 as weight storage types. The experiment with bfloat16 is used as baseline. 

#### SST-2 bf16 training and validation loss
W B Chart 18_09_2026, 19_13_19
W B Chart 18_09_2026, 19_13_45

#### SST-2 Bfp.Bf8 training and validation loss
W B Chart 18_09_2026, 19_14_59
W B Chart 18_09_2026, 19_15_10

#### SST-2 Bfp.Bf4 training and validation loss
W B Chart 21_09_2026, 17_38_27
W B Chart 21_09_2026, 17_38_45

### Checklist
- [x] Main training workflow
- [ ] CPU testing
- [ ] GPU testing
- [x] TT device testing
- [ ] CI testing
    - [ ] Existing tests are passing
    - [ ] Register test to CI
    - [ ] Upload golden files if necessary

### CODE REVIEW

File: blacksmith/datasets/torch/sst2/sst2_utils.py [hunk 1]
```diff
@@ -21,5 +21,7 @@
 LBL2VALUE = {0: "negative", 1: "positive"}
 VALUE2LBL = {"negative": 0, "positive": 1}
 
+# The namespaced id: `datasets` >= 5 rejects bare canonical names, and older releases
+# resolve "glue" to this repo anyway.
```
Comment (mmilosevicTT) [inline blacksmith/datasets/torch/sst2/sst2_utils.py]: Unnecessary change.

File: blacksmith/experiments/torch/llama/crank/lora/single_chip/llama_3_1_8b_alpaca_bfp8.yaml
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/crank/lora/single_chip/llama_3_1_8b_alpaca_bfp8.yaml:1]: This is an ask from all, let's please extend TrainerConfig and not TrainingConfig. This config is really sparse and it will be harder to integrate Trainer for our purpose. In this way we can easily merge configs from experiments to Trainer settings we need.

@umalesTT @agobeljicTT @abogdanovicTT @tstepanovicTT

File: blacksmith/experiments/torch/llama/crank/lora/README.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
+# Llama with LoRA Experiment in tt-crank
+
+This directory contains the code for the Llama with LoRA fine-tuning experiment run through
+tt-crank, the `torch.compile` `tt` backend that ships inside tt-mlir (no tt-xla in the loop).
+It mirrors the [TT-XLA LoRA experiment](../../xla/lora/README.md): same datasets, logger,
+checkpoints and validation cadence, driven by `train_crank.py` instead of `train.py`.
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/crank/lora/README.md]: Let's not mix tt-xla here. README should be fresh.

File: blacksmith/experiments/torch/llama/crank/lora/README.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+export TT_MLIR_HOME=/path/to/tt-mlir
+source env/activate --mlir
+```
+
+The first activation installs the blacksmith-specific packages from `env/mlir_requirements.txt`
+into the tt-mlir venv.
+
+## Training
+
+All configs run on a single P150 (Blackhole) chip.
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/crank/lora/README.md]: This is a bit specific. We might extend in future.

File: blacksmith/experiments/torch/llama/crank/lora/README.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+| P150 | [llama_3_1_8b_alpaca_bfp8.yaml](single_chip/llama_3_1_8b_alpaca_bfp8.yaml) | Alpaca | bfp_bf8 | LoRA |
+
+The configs are short runs capped with `max_steps`, matching the tt-crank experiments they
+were ported from. Remove `max_steps` to train the full epoch(s).
+
+## Data
+
+See the [TT-XLA LoRA README](../../xla/lora/README.md#data) for the SST-2 description. Alpaca
+(`tatsu-lab/alpaca`) is an instruction dataset of 52K instruction/input/output triples; the loader
+splits off 2% of it for validation.
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/crank/lora/README.md]: Not see tt-xla please.

File: blacksmith/experiments/torch/llama/crank/configs.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+from typing import Optional
+
+from pydantic import Field
+
+from blacksmith.experiments.torch.llama.configs import TrainingConfig
+
+
+class CrankTrainingConfig(TrainingConfig):
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/crank/configs.py]: Let's please use Trainer config here and then we can add on that. It will be easier to productise later.

File: blacksmith/experiments/torch/llama/crank/loss.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+# SPDX-FileCopyrightText: (c) 2026 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+import torch
+
+
+def negative_log_likelihood_loss(
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/crank/loss.py]: This should go to utils for sft. This will probably be reused by different models.

File: blacksmith/tools/trainer/trainer.py [hunk 7]
```diff
@@ -57,6 +55,10 @@ def setup(
         if self.reproducibility_manager is None:
             self.reproducibility_manager = ReproducibilityManager(config)
         self.reproducibility_manager.setup()
+        # Imported here, like torch_xla below, so the trainer package loads in environments
+        # without tt-xla (cli.py imports TrainerConfig, which pulls in this module).
```
Comment (mmilosevicTT) [inline blacksmith/tools/trainer/trainer.py:59]: Do we really want to keep tt-xla dependency in `cli.py`? I would not mix trainer changes with this PR. We should definitely first merge cli changes and crank integration and then talk about this.

File: blacksmith/tools/checkpoints_manager.py [hunk 8]
```diff
@@ -142,7 +153,9 @@ def save_checkpoint(
         checkpoint_path = os.path.join(self.checkpoint_dir, checkpoint_name)
 
         trainable_names = {name for name, param in model.named_parameters() if param.requires_grad}
-        state_dict = {name: v for name, v in model.state_dict().items() if name in trainable_names}
+        # Saved on CPU: torch.save cannot serialize tensors on the tt (tt-crank) device, and a
+        # CPU checkpoint loads onto any backend.
```
Comment (mmilosevicTT) [inline blacksmith/tools/checkpoints_manager.py:157]: Is this temporary fix? If it is, please make an issue and link it.

File: blacksmith/experiments/torch/llama/crank/train_crank.py [hunk 9]
```diff
@@ -0,0 +1,659 @@
...
+        raise RuntimeError("weight tying did not survive .to(device) -- the run would train two copies")
+
+    return model, base_model, tied
+
+
+def build_functional_state(model: torch.nn.Module) -> FunctionalState:
+    """Split the module's tensors into the dicts `functional_call` substitutes.
+
+    Trainability follows the `requires_grad` flags peft (or full fine-tuning) left on the
+    module, so `CheckpointManager` sees the same trainable set.
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/crank/train_crank.py]: we have load adapter from https://huggingface.co/docs/hub/en/adapters we should definitely use it if this makes problems.

File: blacksmith/experiments/torch/llama/crank/train_crank.py [hunk 10]
```diff
@@ -0,0 +1,659 @@
...
+    # The module stays in eval mode for the whole run: train()/eval() is baked into the
+    # compiled graph, and Llama has no dropout for train() to enable.
+    model.eval()
+    model.to(device)
+
+    # .to(device) does not preserve weight tying: _has_compatible_shallow_copy_type is False
+    # for a CPU->tt conversion, so nn.Module._apply replaces each Parameter rather than
+    # swapping its .data, and lm_head.weight stops being embed_tokens.weight. Left alone the
+    # run trains two drifting copies of one logical weight with two sets of AdamW moments.
+    # No-op on CPU.
```
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/crank/train_crank.py]: Is this a temporary issue or final workaround?

File: blacksmith/experiments/torch/llama/crank/train_crank.py
Comment (mmilosevicTT) [inline blacksmith/experiments/torch/llama/crank/train_crank.py:1]: Many functions here can be utils.

**Review by mmilosevicTT — COMMENTED** [review]
This PR looks ok to me, but there are a lot of patches that need to wait the tt-crank integration to tt-blacksmith, so we do it in the most compact way.

File: blacksmith/experiments/torch/llama/crank/train_crank.py
Comment (agobeljicTT) [inline blacksmith/experiments/torch/llama/crank/train_crank.py:1]: If we use `/llama/crank/` we can probably name this file just `train.py`

File: blacksmith/tools/trainer/trainer.py (hunk 7 above)
Comment (agobeljicTT) [reply blacksmith/tools/trainer/trainer.py:59]: the way I see it, we can keep this PR with just the experiment (that works locally with crank) with no side customizations
since I am taking care of integration with crank, I'll pick that up
wdythink @tstepanovicTT

File: blacksmith/tools/trainer/trainer.py [hunk 11]
```diff
@@ -114,6 +116,8 @@ def _make_step_loss(self) -> torch.Tensor:
     def _apply_tt_compile_options(self) -> None:
         if not self.config.use_tt:
             return
+        import torch_xla
+
```
Comment (agobeljicTT) [inline blacksmith/tools/trainer/trainer.py:120]: I am personally against conditional imports
I am planning to split utils to crank and xla ones so you would basically just import the crank trainer and you won't need these.

File: CLAUDE.md [hunk 12]
```diff
@@ -11,6 +11,7 @@ Optimized ML training recipes for Tenstorrent hardware using TT-Forge compiler s
 ## Setup & Commands
 ```bash
 source env/activate --xla    # Activate environment (required before ANY work)
+source env/activate --mlir   # tt-crank experiments instead: reuses tt-mlir's venv (needs TT_MLIR_HOME)
```
Comment (agobeljicTT) [inline CLAUDE.md:14]: nice changes really
still this will be done in another PR, for the time being focus on just the experiment (the train_crank.py and configs)

integration will be done afterwards

**Review by agobeljicTT — COMMENTED** [review]
This is cool change. I would just trim the crank integration specifics, that will be a separate PR. Keep them local in order to unblock yourself, but the way I see it, in this PR we should only have the experiment and configs (ofc with any frontend-agnostic things that are needed to enable the experiment)

File: blacksmith/experiments/torch/llama/crank/train_crank.py (hunk 10 above)
Comment (tstepanovicTT) [reply blacksmith/experiments/torch/llama/crank/train_crank.py]: final workaround

File: blacksmith/tools/checkpoints_manager.py (hunk 8 above)
Comment (tstepanovicTT) [reply blacksmith/tools/checkpoints_manager.py:157]: As far as I've found, this is the intended behavior because it keeps a checkpoint loadable on any backend without the source device present.

File: blacksmith/experiments/torch/llama/crank/train_crank.py (hunk 9 above)
Comment (tstepanovicTT) [reply blacksmith/experiments/torch/llama/crank/train_crank.py]: There are no problems as far as I know
